src/Nethermind/Nethermind.Evm/Tracing/GethStyle/Custom/Native/Prestate/NativePrestateTracer.cs

```diff
@@ -0,0 +1,154 @@
+        IWorldState worldState,
+        NativeTracerContext context,
+        GethTraceOptions options) : base(worldState, options)
+    {
+        IsTracingRefunds = true;
+        IsTracingActions = true;
+        IsTracingMemory = true;
+        IsTracingStack = true;
+
+        _prestate = new Dictionary<Address, NativePrestateTracerAccount>();
+
+        if (context.From is not null)
+        {
+            LookupAccount(context.From);
+        }
+        if (context.To is not null)
+        {
+            LookupAccount(context.To);
+        }
+
+        // Look up client coinbase address as well
+        LookupAccount(Address.Zero);
+    }
+
+    protected override GethLikeTxTrace CreateTrace() => new();
+
+    public override GethLikeTxTrace BuildResult()
+    {
+        GethLikeTxTrace result = base.BuildResult();
+
+        result.CustomTracerResult = new GethLikeCustomTrace() { Value = _prestate };
+        return result;
+    }
+
+    public override void StartOperation(int depth, long gas, Instruction opcode, int pc, Address executingAccount, bool isPostMerge = false)
+    {
+        base.StartOperation(depth, gas, opcode, pc, executingAccount, isPostMerge);
+
+        _op = opcode;
+        _executingAccount = executingAccount;
+    }
+
+    public override void SetOperationMemory(TraceMemory memoryTrace)
+    {
+        base.SetOperationMemory(memoryTrace);
+        _memoryTrace = memoryTrace;
+    }
+
+    public override void SetOperationStack(TraceStack stack)
+    {
+        base.SetOperationStack(stack);
+
+        int stackLen = stack.Count;
+        Address address;
+
+        switch (_op)
+        {
+            case Instruction.SLOAD:
+            case Instruction.SSTORE:
+                if (stackLen > 1)
+                {
+                    UInt256 index = stack.PeekUInt256(0);
+                    LookupStorage(_executingAccount!, index);
+                }
+                break;
+            case Instruction.EXTCODECOPY:
+            case Instruction.EXTCODEHASH:
+            case Instruction.EXTCODESIZE:
+            case Instruction.BALANCE:
+            case Instruction.SELFDESTRUCT:
+                if (stackLen >= 1)
+                {
+                    address = stack.Peek(0).ToHexString(true, true).ToAddress();
+                    LookupAccount(address);
+                }
+                break;
+            case Instruction.DELEGATECALL:
+            case Instruction.CALL:
+            case Instruction.STATICCALL:
+            case Instruction.CALLCODE:
+                if (stackLen >= 5)
+                {
+                    address = stack.Peek(1).ToHexString(true, true).ToAddress();
+                    LookupAccount(address);
+                }
+                break;
+            case Instruction.CREATE2:
+                if (stackLen >= 4)
+                {
+                    int offset = int.Parse(stack.Peek(1));
+                    int length = int.Parse(stack.Peek(2));
```

Comment:
possible exceptions, stack can have UInt256

Reply:
Is this fine with throwing the default `ThrowOverflowOrFormatException` exception or should I do a `TryParse` and throw a different exception here in the error case?

Reply:
On second thought, would using `stack.Peek(1).ReadEthInt32()` be preferable here?

Reply:
I think you shouldn't fail execution here. If I am correct, this is pre-actual VM execution for this instruction that will fail with `OutOfGas`, which is preferable to tracer throwing exceptions.